File: Easy/283_Move_Zeroes.java
 class Solution {
    public void moveZeroes(int[] nums) {

        //indice j will keep track my non-zero element position
        int j = 0;
        //indice i will traverse down the array looking for zero element
        for(int i = 0; i < nums.length; i++){
            if(nums[i] != 0){
                nums[j] = nums[i];
                j++;
                //printArray(nums);
            }
        }
        // array.length - j == the no. of 0 occurences , hence we just fill up the back with 0s
        while(j < nums.length){
            nums[j] = 0;
            j++;
        }
    }
    public static void printArray(int[] array){
        System.out.printf("[");
        for(int i = 0; i < array.length; i++){

            System.out.printf("%d",array[i]);
            if(i < array.length - 1)
                System.out.printf(",");
        }
        System.out.println("]");
    }
}

File: LeetCode 75 Study Plan/283_Move_Zeroes.js
/**
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var moveZeroes = function(nums) {
    /*
        Index j keep track of the non-zero element positios, maintaining the 
        relative order of the non-zero elements.
    */
    let j = 0; 

    for(let i = 0; i < nums.length; i++){
        if(nums[i] != 0){
            nums[j] = nums[i];
            j++;
        }
    }

    /** 
        array.length - j == the no. of 0 occurences, hence we just fill up 
        the back of the arrayswith 0s.
    */
    while(j < nums.length){
        nums[j] = 0;
        j++;
    }
};
